services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/RouterSubnetsCache.java

```diff
@@ -20,89 +20,62 @@
     @Autowired
     public RouterSubnetsCache(CacheFactory cacheFactory) {
-        this.routerSubnetsCache = cacheFactory.getCache(InternalSubnets.class);
+        this.cacheFactory = cacheFactory;
+        this.routerSubnetsCache = cacheFactory.getCache(InternalSubnetRouterMap.class);
     }
 
     @DurationStatistics
-    public InternalSubnets getRouterSubnets(String routerId) throws CacheException {
-        return routerSubnetsCache.get(routerId);
-    }
+    public Collection<InternalSubnetRouterMap> getRouterSubnets(String routerId) throws CacheException {
+        Map<String, Object[]> queryParams = new HashMap<>();
+        Object[] values = new Object[1];
+        values[0] = routerId;
+        queryParams.put("routerId", values);
+        return routerSubnetsCache.getAll(queryParams).values();
 
-    @DurationStatistics
-    public Map<String, InternalSubnets> getAllSubnets() throws CacheException {
-        return routerSubnetsCache.getAll();
-    }
-
-    @DurationStatistics
-    public Map<String, InternalSubnets> getAllSubnets(Map<String, Object[]> queryParams) throws CacheException {
-        return routerSubnetsCache.getAll(queryParams);
     }
 
     @DurationStatistics
-    public synchronized void addVpcSubnets(InternalSubnets subnets) throws CacheException {
-        routerSubnetsCache.put(subnets.getRouterId(), subnets);
+    public Collection<InternalSubnetRouterMap> updateVpcSubnets(NetworkConfiguration networkConfig) throws CacheException {
+        Map<String, InternalSubnetRouterMap> internalSubnetsMap =  networkConfig
+                .getInternalRouterInfos()
+                .stream()
+                .filter(routerInfo -> routerInfo.getRouterConfiguration().getSubnetRoutingTables().size() > 0)
+                .flatMap(routerInfo -> routerInfo.getRouterConfiguration().getSubnetRoutingTables()
+                        .stream()
+                        .map(routingTable -> new InternalSubnetRouterMap(routerInfo.getRouterConfiguration().getId()
+                                , routingTable.getSubnetId())))
+                .collect(Collectors.toMap(routerInfo -> routerInfo.getRouterId() + cacheFactory.KEY_DELIMITER + routerInfo.getSubnetId(), Function.identity()));
+        routerSubnetsCache.putAll(internalSubnetsMap);
+        return internalSubnetsMap.values();
     }
 
     @DurationStatistics
-    public void updateVpcSubnets(InternalSubnets subnets) throws CacheException {
-        routerSubnetsCache.put(subnets.getRouterId(), subnets);
-    }
-
-    @DurationStatistics
-    public void deleteVpcGatewayInfo(String routerId) throws CacheException {
-        routerSubnetsCache.remove(routerId);
-    }
-
-    @Override
-    public InternalSubnets findItem(String id) throws CacheException {
-        return routerSubnetsCache.get(id);
+    public void deleteVpcGatewayInfo(String routerId, String subnetId) throws CacheException {
+        routerSubnetsCache.remove(routerId + cacheFactory.KEY_DELIMITER + subnetId);
```

Comment:
We have some key change in the key-value store to improve performance in multiple caches in this PR. 

Since DPM design doc is quite outdated, we should update the design doc by putting major design points and gathering optimization techniques applied to DPM cache. @DavidLiu506 @cj-chung

Reply:
Agree, we need to rewrite the DPM design doc and including the new cache design into the doc.

Reply:
Will update doc.

Reply:
Thanks. Create an issue https://github.com/futurewei-cloud/alcor/issues/715 for tracking.

Reply:
Thanks!